File: java-vfs/src/main/java/emg/java/vfs/commands/Rm.java
package emg.java.vfs.commands;

import emg.java.vfs.files.Directory;
import emg.java.vfs.filesystem.State;

import java.util.Arrays;
import java.util.List;

import static emg.java.vfs.extensions.ListExtensions.head;
import static emg.java.vfs.extensions.ListExtensions.tail;

public class Rm extends Command {

    private final String name;

    public Rm(String name) {
        this.name = name;
    }

    @Override
    public State apply(State state) {
        // 1 get working dir
        var wd = state.wd();
        // 2 get absolute path
        var absolutePath = "";
        if (name.startsWith(Directory.SEPARATOR)) {
            absolutePath = name;
        } else if (wd.isRoot()) {
            absolutePath = wd.path() + name;
        } else {
            absolutePath = wd.path() + Directory.SEPARATOR + name;
        }
        // 3 do some checks
        if (Directory.ROOT_PATH.equals(absolutePath)) {
            return state.setMessage("Cannot drop Root");
        } else {
            return doRm(state, absolutePath);
        }
    }

    private State doRm(State state, String path) {
        // 4 find the entry to remove
        // 5 update the structure
        var tokens = Arrays.asList(path.substring(1).split(Directory.SEPARATOR));
        var newRoot = rmHelper(state.root(), tokens);
        if (newRoot == state.root()) {
            return state.setMessage(path + ": no such file or directory");
        } else {
            return State.apply(newRoot, newRoot.findDesendant(state.wd().path().substring(1)));
        }
    }

    private Directory rmHelper(Directory currentDirectory, List<String> path) {
        if (path.isEmpty()) {
            return currentDirectory;
        } else if (tail(path).isEmpty()) {
            return currentDirectory.removeEntry(head(path));
        } else {
            var nextDirectory = currentDirectory.findEntry(head(path));
            if (!nextDirectory.isDirectory()) {
                return currentDirectory;
            } else {
                var newNextDirectory = rmHelper(nextDirectory.asDirectory(), tail(path));
                if (newNextDirectory == nextDirectory) {
                    return currentDirectory;
                } else {
                    return currentDirectory.replaceEntry(head(path), newNextDirectory);
                }
            }
        }
    }
}

File: kotlin-vfs/src/main/kotlin/emg/kotlin/vfs/commands/Rm.kt
package emg.kotlin.vfs.commands

import emg.kotlin.vfs.extensions.head
import emg.kotlin.vfs.extensions.tail
import emg.kotlin.vfs.files.Directory
import emg.kotlin.vfs.filesystem.State

class Rm(val name: String) : Command() {
    override fun apply(state: State): State {
        // 1 get working dir
        val wd = state.wd
        // 2 get absolute path
        val absolutePath = when {
            name.startsWith(Directory.SEPARATOR) -> name
            wd.isRoot() -> wd.path() + name
            else -> wd.path() + Directory.SEPARATOR + name
        }
        // 3 do some checks
        return if (Directory.ROOT_PATH == absolutePath)
            state.setMessage("Cannot drop Root")
        else
            doRm(state, absolutePath)
    }


    private fun doRm(state: State, path: String): State {
        fun rmHelper(currentDirectory: Directory, path: List<String>): Directory {
            return if (path.isEmpty()) currentDirectory
            else if (path.tail().isEmpty()) currentDirectory.removeEntry(path.head())
            else {
                val nextDirectory = currentDirectory.findEntry(path.head())
                if (!nextDirectory!!.isDirectory()) currentDirectory
                else {
                    val newNextDirectory = rmHelper(nextDirectory.asDirectory(), path.tail())
                    if (newNextDirectory == nextDirectory) currentDirectory
                    else currentDirectory.replaceEntry(path.head(), newNextDirectory)
                }
            }
        }

        // 4 find the entry to remove
        // 5 update the structure
        val tokens = path.substring(1).split(Directory.SEPARATOR).toList()
        val newRoot: Directory = rmHelper(state.root, tokens)
        return if (newRoot == state.root)
            state.setMessage("$path: no such file or directory")
        else State.apply(newRoot, newRoot.findDesendant(state.wd.path().substring(1)))

    }
}
